### PR TITLE
Manifest file generation for Component governance code improvement

### PR DESCRIPTION
**Pull request by rijulluman-msft**
Will merge this when my existing changes are in LKG, so I can uncomment the generate argument in bxl.ps1
[AB#1580270](https://dev.azure.com/mseng/708e929f-6bd5-415a-8daf-25b1dac08dd8/_workitems/edit/1580270)

### CODE REVIEW

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 1]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
...
-                    // Validation of existing cgmainfest.json results in failure due to mismatch. Should fail the build in this case.
-                    try
+                    if (    Configuration.FrontEnd.GenerateCgManifestForNugets.IsValid
+                        &&  Configuration.FrontEnd.ValidateCgManifestForNugets.Equals(Configuration.FrontEnd.GenerateCgManifestForNugets))
                     {
-                        existingCgManifest = File.ReadAllText(Configuration.FrontEnd.ValidateCgManifestForNugets.ToString(Context.PathTable));
-                        FrontEndHost.Engine.RecordFrontEndFile(
-                            Configuration.FrontEnd.ValidateCgManifestForNugets,
-                            CGManifestResolverName);
+                        // Skip validation since the newly generated file will always be valid
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]: let's not have empty `if` body and merge this condition with the condition above

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 2]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
             {
                 var cgManfiestGenerator = new NugetCgManifestGenerator(Context);
                 string generatedCgManifest = cgManfiestGenerator.GenerateCgManifestForPackages(maybePackages.Result);
-                string existingCgManifest = "INVALID";
 
-                if ( !Configuration.FrontEnd.GenerateCgManifestForNugets.IsValid &&
-                      Configuration.FrontEnd.ValidateCgManifestForNugets.IsValid )
+                if ( Configuration.FrontEnd.ValidateCgManifestForNugets.IsValid )
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]:
```suggestion
                if (Configuration.FrontEnd.ValidateCgManifestForNugets.IsValid && 
                   !Configuration.FrontEnd.ValidateCgManifestForNugets.Equals(Configuration.FrontEnd.GenerateCgManifestForNugets))
```

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 3]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
...
-                        {
-                            Logger.ReportComponentGovernanceGenerationError(Context.LoggingContext, "Could not write Component Governance Manifest file to disk\n" + e.ToString());
-                            return false;
-                        }
-                    }
-                }
+        private bool ValidateCgManifestFile(string generatedCgManifest)
+        {
+            // Validation of existing cgmainfest.json results in failure due to mismatch. Should fail the build in this case.
+            string existingCgManifest;
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]:
```suggestion
```

Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]:
```suggestion
```

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 4]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
...
+            try
+            {
+                existingCgManifest = File.ReadAllText(Configuration.FrontEnd.ValidateCgManifestForNugets.ToString(Context.PathTable));
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]:
```suggestion
                string existingCgManifest = File.ReadAllText(Configuration.FrontEnd.ValidateCgManifestForNugets.ToString(Context.PathTable));
```

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 5]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
...
+                Logger.ReportComponentGovernanceValidationError(Context.LoggingContext, "Cannot read Component Governance Manifest file from disk\n" + e.ToString());
+                return false;
+            }
+            catch (FileNotFoundException e)
+            {
+                Logger.ReportComponentGovernanceValidationError(Context.LoggingContext, "Cannot read Component Governance Manifest file from disk\n" + e.ToString());
+                return false;
+            }
+
+            if (!NugetCgManifestGenerator.CompareForEquality(generatedCgManifest, existingCgManifest))
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]: move the code from here until the end of the method inside the try/catch

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs [hunk 6]
```diff
@@ -80,79 +80,85 @@ public override async Task<bool> InitResolverAsync(IResolverSettings resolverSet
...
             }
-
-            m_resolverState = State.ResolverInitialized;
+            catch (BuildXLException e)
+            {
+                Logger.ReportComponentGovernanceGenerationError(Context.LoggingContext, "Could not write Component Governance Manifest file to disk\n" + e.ToString());
+                return false;
+            }
 
             return true;
```
Comment (almiliMSFT) [inline Public/Src/FrontEnd/Nuget/NugetResolver.cs]: move inside try/catch

File: Public/Src/FrontEnd/Nuget/NugetResolver.cs (hunk 2 above)
Comment (rijulluman-msft) [reply Public/Src/FrontEnd/Nuget/NugetResolver.cs]: Will this also check validity of Configuration.FrontEnd.GenerateCgManifestForNugets ?

Comment (almiliMSFT) [reply Public/Src/FrontEnd/Nuget/NugetResolver.cs]: No but it doesn't need to.  If ValidateCgManifest is valid and GenerateCgManifest is not, `Equals` will return false so this `if` condition will be satisfied

**Pull request merged**